include/wil/win32_helpers.h

```diff
@@ -531,6 +531,105 @@ namespace wil
             FreeLibraryAndExitThread(thisModule, 0);
         });
     }
+
+    struct module_reference_for_new_thread
+    {
+        bool free_on_exit = false;
+        HMODULE module_ref = nullptr;
+
+        module_reference_for_new_thread()
+        {
+            FAIL_FAST_IF(!::GetModuleHandleExW(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS, L"", &module_ref));
+        }
+
+        ~module_reference_for_new_thread()
+        {
+            reset();
+        }
+
+        module_reference_for_new_thread(module_reference_for_new_thread&& other)
+        {
+            wistd::exchange(free_on_exit, other.free_on_exit);
+            wistd::exchange(module_ref, other.module_ref);
+        }
+
+        module_reference_for_new_thread& operator=(module_reference_for_new_thread&& other)
+        {
+            if (this != wistd::addressof(other))
+            {
+                reset();

```

Comment:
nit: Assuming you change to `swap`, calling `reset` here is not necessary. Moved-from objects are specified as being in a "valid but unspecified state", so giving the moved-from object the current object's resources to free in its destructor is a relatively common pattern